python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,1111 @@
+        # candidate for. We break the loop if we haven't found enough.
+        n_fail = 0
+        # Loop over the source pairs.
+        for src_idx in xrange(1, len(src_dist_array)):
+            if n_fail > len(src_dist_array) - (n_match - 1):
+                break
+
+            # Given our length tolerance we can use it to compute a tolernace
+            # on the angle between our spoke.
+            src_sin_tol = (max_dist_rad /
+                           (src_dist_array[src_idx] + max_dist_rad))
+            # Test if the small angle approximation will still hold. This is
+            # defined as when sin(theta) ~= theta to within 0.1% of each other.
+            # This also implicitly sets a minimum spoke length that we can use.
+            if src_sin_tol > 0.0447:
+                n_fail += 1
+                continue
+
+            # Find the reference pairs that include our candiates pattern
+            # center and sort them in increasing delta
+            ref_dist_idx_array = self._find_candidate_reference_pairs(
+                src_dist_array[src_idx], ref_dist_array, max_dist_rad)
+
+            # Test the spokes and return the id of the reference object.
+            # Return None if no match is found.
+            ref_id = self._test_spoke(
+                src_ctr, src_delta_array[src_idx], src_dist_array[src_idx],
+                src_delta_array[0], src_dist_array[0], ref_ctr, ref_ctr_id,
+                ref_delta, ref_dist, ref_dist_idx_array, ref_delta_array,
+                ref_dist_array,
+                ref_id_array, src_sin_tol)
+            if ref_id is None:
+                n_fail += 1
+                continue
+
+            # Append the successful indicies to our list. The src_idx needs
+            # an extra iteration to skip the first and second source objects.
+            ref_list.append(ref_id)
+            src_list.append(src_idx + 1)
+            # If we found enough reference objects we can return early. This is
+            # n_match - 2 as we already have 2 source objects matched into the
+            # reference data.
+            if len(ref_list) >= n_match - 2:
+                break
+        return pipeBase.Struct(
+            ref_spoke_list=ref_list,
+            src_spoke_list=src_list,)
+
+    # TODO: Possibly clean up the arguments here by pre computing
+    # the source dot and cross products of the opening angle in the
+    # previous method.
+    def _test_spoke(self, src_ctr, src_delta, src_dist, src_ctr_delta,
+                    src_ctr_dist, ref_ctr, ref_ctr_id, ref_delta, ref_dist,
+                    ref_dist_idx_array, ref_delta_array, ref_dist_array,
+                    ref_id_array, src_sin_tol):
+        """Test the opening angle between the first spoke of our pattern
+        for the soruce object against the reference object. This method makes
+        heavy use of the small angle approximation assumption to perform the
+        comparison.
```

Comment:
Separate brief description from rest of description

"soruce"

Consider eliding "assumption" to give: "heavy use of the small angle approximation to perform the comparison"